wurst/event/DamageQueue.wurst

```diff
@@ -0,0 +1,108 @@
+			// Make damage resonance
+		addDamageListener() d ->
+			Log.info("End " + (d castTo int).toString())
+
+		// To demonstrate why is this even important
+		addDamageModifier() d ->
+			if d.getAmount() < d.getTarget().getMaxHP() / 2
+				d.preventDeath()
+				// Some tough guy's ability, who can't die from weak attacks
+
+	Then somewhere happens this:
+	unit2 has 101 hp out of 101.
+	unit1.damageTarget(unit2, 75, 2)
+	(Causing then 50 and 25 damage)
+	
+	Logs:
+	Start 1
+	Start 2
+	Start 3
+	End 3
+	*unit 2 is hit with 25 damage, 76 left*
+	End 2
+	*unit 2 is hit with 50 damage, 26 left*
+	End 1
+	*unit 2 is hit with 75 (26 actual) damage, unit 2 dies.*
+
+	If you replace .damageTarget with .damageTargetQueued, you'll get this
+	result with in same test case:
+
+	Start 1
+	End 1
+	*unit 2 is hit with 75 damage, 26 left*
+	Start 2
+	End 2
+	*unit 2 is hit with 50 (25.594 actual) damage, 0.406 left*
+	Start 3
+	End 3
+	*unit 2 is hit with 25 (0 actual) damage, unit 2 lives*
+
+	So the results in this test are different if we take order into consideration.
+
+	This is proved for war3 version 1.30.
+	As you can see, in second case damage processing follows the order given
+*/
+
+@configurable constant RECURSION_TRESHOLD = 128
+
+/** This is a replacement for .damageTarget() in case of nested damage call that
+	delays new damage event to run after previous damage was dealt.
+	
+	More info on that in DamageQueue package. */
+public function unit.damageTargetQueued(unit target, real amount, bool attack, bool ranged, attacktype attackType, damagetype damageType, weapontype weaponType, int data) returns Damage
+	return new Damage(this, target, amount, attack, ranged, attackType, damageType, weaponType, data)..saveInstance()..modify()..dealQueued()
+
+/** This is a replacement for .damageTarget() in case of nested damage call that
+	delays new damage event to run after previous damage was dealt.
+	
+	More info on that in DamageQueue package. */
+public function unit.damageTargetQueued(unit target, real amount, attacktype attackType, int data) returns Damage
+	return this.damageTargetQueued(target, amount, false, false, attackType, DAMAGE_TYPE_UNIVERSAL, WEAPON_TYPE_WHOKNOWS, data)
+
+/** This is a replacement for .damageTarget() in case of nested damage call that
+	delays new damage event to run after previous damage was dealt.
+	
+	More info on that in DamageQueue package. */
+public function unit.damageTargetQueued(unit target, real amount, int data) returns Damage
+	return this.damageTargetQueued(target, amount, false, false, ATTACK_TYPE_CHAOS, DAMAGE_TYPE_UNIVERSAL, WEAPON_TYPE_WHOKNOWS, data)
+
+let queue = new LinkedList<Damage>
+/** This is a replacement for .deal() in case of nested damage call that
+	delays new damage event to run after previous damage was dealt.
+	
+	More info on that in DamageQueue package. */
+public function Damage.dealQueued()
+	if queue.size() < 128
+		if queue.isEmpty()
+			getTimer()..start(0., function queueFlush)
```

Comment:
There's also a part of the event we cannot control - health withdrawal. So there's no way to call Unitdamagetarget inside damage callback and then make it happen after that damage. We can only delay the call.

There's other way to catch an event when the damage is dealt - the more correct one - unit life changed event, but that requires rebuilding a trigger once a damage event which is too much overhead for me.

If you are not consistently using Queued damage, then there's no guarantees, it's right (A thing to mention in hotdoc, thanks). So for example AoE damage calling default UDT would have first all the aoe damage and only then all damage caused by it's callbacks.

Reply:
`unit.damageTarget` is not a part of this system, so I can't change its behavior. There are some wrappers that overload it with an additional field, and they can use this method of delaying the damage. Even if I would remove/rewrite native `unit.damageTarget`, there's `UnitDamageTarget` native which is not prohibited to use and I cannot change how it is called.

At least since this system is an optional module, it should work around default way to deal damage and not replace it.

Reply:
We could put this check inside of the handler, which would have the same effect as placing it into the `unit.damageTarget()` calls that exist outside this library. 

Any calls to any damage dealing natives should trigger your handler, which could then look at the given global state and place the processing of the event into the queue or start a new queue, depending on the state.

Reply:
I could write a little mock of a system like this, if you are interested in a more concrete demonstration of what I'm talking about.

Reply:
Ok, so if we are working only in the handler part, then we have no way to stop/delay event from happening. And I'm talking about what game processes - the target recieves the damage and losses hp and maybe dies with credits to damage source.
We should delay unitdamagetarget call, which is impossible if it's already called.

Or your proposal is to make it a 0 damage event and then re-call it?

Reply:
Unless you try to damage more than 3+ enemies with a very small interval (0.03 and below), using a trigger should be fine for detecting the after damage event.

When dealing with modified damage, (before the existence of BlzSetEventDamage), try to set the health to a bare minimum such that the original damage deals the killing blow, if it is a killing blow.

Now, here's another potential overhead, asserting how much health would be lost. For small values, simple arithmetic should suffice. But when dealing with large values, you can simulate how much health is lost via SetWidgetLife (just make sure to not to kill the unit this way).

If there is no difference between the former health and the latter (result from SetWidgetLife assertion or damage was initially 0.00), fire the After Damage trigger immediately. Otherwise, leave it be.

Reply:
Any AOE damage or crowded battle are bounded to use this system. Too bad for stdlib.
Difference is the order in which things happen. Currently this system leaves proper damage amount for next damage handlers from ddetection, and it should not be changed, so emulating damage dealing is not an option too.
Actually any workaround seems to be too messy, excluding proper order from the beginning.